bin/oneoff/export_pardot_contacts.rb

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env ruby
+
+# Export Pardot contacts from the main contact_rollups table to CSV files
+# to manually upload them to Pardot.
+#
+# This is a workaround for when we can not upload all the changes to Pardot due to
+# daily API limit of 25k queries/day.
+#
+# Output files are pardot_new_contacts.csv and pardot_updated_contacts.csv.
+# Maximum number of rows for each file is 1 million. If there are more than 1m rows,
+# it will write to additional files, e.g. pardot_updated_contacts_1.csv.
+
+require_relative '../../deployment'
+require 'cdo/pegasus'
+require 'cdo/pardot'
+
+MAX_EXECUTION_TIME = 1_800_000
+MAX_EXECUTION_TIME_SEC = MAX_EXECUTION_TIME / 1000
+
+PEGASUS_DB_READER = sequel_connect(
+  CDO.pegasus_db_reader,
+  CDO.pegasus_db_reader,
+  query_timeout: MAX_EXECUTION_TIME_SEC
+)
+
+def all_prospect_fields
+  pardot_fields = Pardot::MYSQL_TO_PARDOT_MAP.values.pluck(:field)
+  special_fields = %w(opted_out is_do_not_email db_Opt_In db_Has_Teacher_Account db_Imported)
+  pardot_fields + special_fields
+end
+
+def convert_row_to_prospect(row)
+  prospect = {}
+  Pardot::MYSQL_TO_PARDOT_MAP.each do |mysql_key, pardot_info|
+    db_value = row[mysql_key]
+    next unless db_value.present?
+
+    if pardot_info[:multi]
+      # For multi data fields (multi-select,etc.), we set value names as
+      # [field_name]_0, [field_name]_1, etc.
+      values = db_value.split(",")
+      values.each_with_index do |value, index|
+        prospect["#{pardot_info[:field]}_#{index}"] = value
+      end
```

Comment:
Are the multi values ending up in your final file correctly?  I see you pushing these values with numbered suffixes into the "prospect" hash, but later when you pluck values back out of the hash into the CSV you're using the original column list with no numbered suffixes - so I think the values for the multi columns get lost.

Reply:
Great catch. Investigating...

Reply:
This is tricky because we won't know what columns to write to CSV file (which is the very first line in the file) until we go through all the rows in the dataset.

Fix by doing another pass through the dataset to build column list. This is not optimal, but does the job.